### PR TITLE
ICU-22818 Fix memory leak during error in messageformat2_data_model.cpp

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22818
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [X] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/i18n/messageformat2_data_model.cpp [hunk 1]
```diff
@@ -1094,6 +1094,8 @@ MFDataModel::MFDataModel(const MFDataModel::Builder& builder, UErrorCode& errorC
         Variant* variants = copyVectorToArray<Variant>(*builder.variants, errorCode);
         Expression* selectors = copyVectorToArray<Expression>(*builder.selectors, errorCode);
         if (U_FAILURE(errorCode)) {
+            delete variants;
```
Comment (aheninger) [inline icu4c/source/i18n/messageformat2_data_model.cpp:1097]: `copyVectorToArray` is doing an array new, so at a minimum the deletes here need to be array deletes.

Even better would be to use `LocalArray`

**Review by aheninger — DISMISSED** [review]

File: icu4c/source/i18n/messageformat2_data_model.cpp (hunk 1 above)
Comment (FrankYFTang) [reply icu4c/source/i18n/messageformat2_data_model.cpp]: PTAL

File: icu4c/source/i18n/messageformat2_data_model.cpp [hunk 2]
```diff
@@ -1091,13 +1091,13 @@ MFDataModel::MFDataModel(const MFDataModel::Builder& builder, UErrorCode& errorC
         U_ASSERT(builder.selectors != nullptr);
         int32_t numVariants = builder.variants->size();
         int32_t numSelectors = builder.selectors->size();
-        Variant* variants = copyVectorToArray<Variant>(*builder.variants, errorCode);
-        Expression* selectors = copyVectorToArray<Expression>(*builder.selectors, errorCode);
+        LocalArray<Variant> variants(copyVectorToArray<Variant>(*builder.variants, errorCode), errorCode);
+        LocalArray<Expression> selectors(copyVectorToArray<Expression>(*builder.selectors, errorCode), errorCode);
```
Comment (richgillam) [inline icu4c/source/i18n/messageformat2_data_model.cpp:1095]: Do these arrays own their contents?  If so, you need to account for that.

Comment (FrankYFTang) [reply icu4c/source/i18n/messageformat2_data_model.cpp:1095]: I did.   I call orphan() to release the ownership

Comment (richgillam) [reply icu4c/source/i18n/messageformat2_data_model.cpp:1095]: orphan() releases ownership of the arrays themselves.  I was asking about the _elements_ of the arrays.  I don't know what `Variant` and `Expression` are.  But if I'm reading this right, they're inline structs, not pointers, and so there isn't an ownership issue here and I asked the question in error.  Sorry about that.

**Review by richgillam — APPROVED** [review]

**Pull request merged**